### ISSUE TITLE
Default training not converging

### ISSUE THREAD

**Issue by skanderm** [astorfi/3D-convolutional-speaker-recognition#3]
Running just the `train_softmax.py` command in the example `run.sh` script with the sample data doesn't seem to converge, even at 50 epochs.

Command:

`python -u ./code/1-development/train_softmax.py --num_epochs=50 --batch_size=3 --development_dataset_path=data/development_sample_dataset_speaker.hdf5 --train_dir=results/TRAIN_CNN_3D/train_logs`

Output:

![image](https://user-images.githubusercontent.com/997770/30943708-9cea6692-a3a7-11e7-94ac-220f7702fe78.png)

Loss:

![image](https://user-images.githubusercontent.com/997770/30943715-af5d96dc-a3a7-11e7-98d3-b16c6975fd71.png)

Learning rate:

![image](https://user-images.githubusercontent.com/997770/30943749-e62d42de-a3a7-11e7-9298-b8514e979b83.png)

**Comment by astorfi** [comment]
@skanderm The default training is just a running example on a randomly generated dataset for a demonstration of the training process and how all the parts are connected together! You should try your own dataset in a correct format and the details are available in the associated paper!

**Comment by skanderm** [comment]
Thanks Sina, that makes sense. I'd hoped the sample data would provide a proportional/converging example, but it's good to know it's not supposed to.

A few more questions:

1. According to the paper, the data set consists of 1083 speakers. Did the development model use all 1083 speakers?
2. How many cubes did you use per speaker for development (the x in `(x, 80, 40, ζ=20)`)?
3. How many epochs would you estimate was required to converge? I don't see that in the paper.
4. How many speakers were enrolled and evaluated? All 1083 again?
5. The dimensionality of the data for enrollment_evaluation is `(x, 80, 40, ζ=1)` while the development is `(x, 80, 40, ζ=20)`. Is that correct/representative?
6. Related to 2, how many cubes (`x`) for enrollment?

Thank you!

**Comment by astorfi** [comment]
First of all, you said, "Given the dimensionality of the sample data, represented by (n*x, 80, 40, ζ), with n being the number of speakers and x being cubes per speaker". My question is how did you come up with n*x? Did I mention anywhere as this? Because the first dimension is simply the sample index and each sample is a cube. Although your understanding is correct and n*x can be (not always) the number of samples.

1. N has nothing to do with the 1083 or the number of speakers. It is the number of generated samples and can be related to the number of speakers. In the dataset, there are 1083 speaker but only around 600 of them are used in the experiments! (Please refer to paper for further details)

2. Same as above.

3. It is actually database related. For us, at least 50 epochs.

4. In the enrollment, the ζ is 20 as well which is the main idea of the paper. Creating a bridge between development and enrollment. for the evaluation, ζ = 1 but we create copies to has ζ = 20 for the same above reasoning. (Please refer to paper for further details)

**Comment by skanderm** [comment]
Thanks @astorfi. No, you didn't mention `n*x` anywhere in the paper. I figured `n*x` was a good summary of what the sample data contained. 

I'm trying to get a sense of the volume and breakdown of data to achieve 86%+ AUC. I'm not seeing 600 speakers mentioned in the paper, just the 1083 speakers:

> The dataset that has been used for our experiments is the WVU-Multimodal 2013 dataset [19]. The audio part of WVU-Multimodal dataset consists of up to 4 sessions of interviews for each of the 1083 different speakers. The WVU-Multimodal dataset includes different modalities of data collected over a period from 2013 to 2015. The audio part of data consists of both scripted and unscripted voice samples. For the scripted samples, the participants read a fixed sample of text. For the unscripted samples, the participants answer interview questions that require conversational responses. We only use the scripted audio samples, as only the voice of the subject of interest is present in the sample.

So: How many samples did you require, roughly, per speaker to achieve those results? I have roughly 100 speakers in my data set and I'm trying to figure out how to split them up. 

Thank again!

**Comment by astorfi** [comment]
Section VI

> The output of the last layer (FC5) will be fed
> to the softmax layer which has the cardinality of N = 511,
> where N is the number of speakers in the development phase.
> For the enrollment and evaluation stages, 100 subjects have
> been used and the speaker utterances are split into two equal
> parts for two aforementioned phases.

However, I accept it could be explained in a better way in the paper.

**Comment by skanderm** [comment]
Oops, I might have had an old version of the paper: 

![image](https://user-images.githubusercontent.com/997770/31398061-37d32ed0-add7-11e7-96da-2f31adda2f30.png)

I'll have a look at the newest version and re-open if there's an issue. Thanks for clarifying!
